packages/block-library/src/table/state.js

```diff
@@ -26,105 +28,186 @@ export function createTable( {
-	section,
+export function getFirstRow( state ) {
+	if ( ! isEmptyTableSection( state.head ) ) {
+		return state.head[ 0 ];
+	}
+	if ( ! isEmptyTableSection( state.body ) ) {
+		return state.body[ 0 ];
+	}
+	if ( ! isEmptyTableSection( state.foot ) ) {
+		return state.foot[ 0 ];
+	}
+}
+
+/**
+ * Gets an attribute for a cell.
+ *
+ * @param {Object} state 			   	 Current table state.
+ * @param {string} options.sectionName	 Section of the cell to update.
+ * @param {number} options.rowIndex    	 Row index of the cell to update.
+ * @param {number} options.columnIndex 	 Column index of the cell to update.
+ * @param {number} options.attributeName The name of the attribute to get the value of.
+ *
+ * @return {*} The attribute value.
+ */
+export function getCellAttribute( state, {
+	sectionName,
 	rowIndex,
 	columnIndex,
-	content,
+	attributeName,
 } ) {
-	return {
-		[ section ]: state[ section ].map( ( row, currentRowIndex ) => {
-			if ( currentRowIndex !== rowIndex ) {
+	return get( state, [ sectionName, rowIndex, 'cells', columnIndex, attributeName ] );
+}
+
+/**
+ * Returns updated cell attributes after applying the `updateCell` function to the selection.
+ *
+ * @param {Object}   state      The block attributes.
+ * @param {Object}   selection  The selection of cells to update.
+ * @param {Function} updateCell A function to update the selected cell attributes.
+ *
+ * @return {Object} New table state including the updated cells.
+ */
+export function updateSelectedCell( state, selection, updateCell ) {
+	if ( ! selection ) {
+		return state;
+	}
+
+	const tableSections = pick( state, [ 'head', 'body', 'foot' ] );
+	const {
+		sectionName: selectionSectionName,
+		rowIndex: selectionRowIndex,
+	} = selection;
+
+	return mapValues( tableSections, ( section, sectionName ) => {
+		if ( selectionSectionName && selectionSectionName !== sectionName ) {
+			return section;
+		}
+
+		return section.map( ( row, rowIndex ) => {
+			if ( selectionRowIndex && selectionRowIndex !== rowIndex ) {
 				return row;
 			}
 
 			return {
-				cells: row.cells.map( ( cell, currentColumnIndex ) => {
-					if ( currentColumnIndex !== columnIndex ) {
-						return cell;
+				cells: row.cells.map( ( cellAttributes, columnIndex ) => {
+					const cellLocation = {
+						sectionName,
+						columnIndex,
+						rowIndex,
+					};
+
+					if ( ! isCellSelected( cellLocation, selection ) ) {
+						return cellAttributes;
 					}
 
-					return {
-						...cell,
-						content,
-					};
+					return updateCell( cellAttributes );
 				} ),
 			};
-		} ),
-	};
+		} );
+	} );
+}
+
+/**
+ * Returns whether the cell at `cellLocation` is included in the selection `selection`.
+ *
+ * @param {Object} cellLocation An object containing cell location properties.
+ * @param {Object} selection    An object containing selection properties.
+ *
+ * @return {boolean} True if the cell is selected, false otherwise.
+ */
+export function isCellSelected( cellLocation, selection ) {
```

Comment:
Why not have `isCellSelected` and `isColumnSelected`?

Reply:
`isColumnSelected` would only be used in one place, so it seems unnecessary at this stage to have two separate functions.

I'm exploring this more in #16493, where I do have separate functions for different types of selections as well as an `isCellSelected` for easily checking whether a given cell is in any type of selection:
https://github.com/WordPress/gutenberg/blob/ec1aa436ec4e4ce641799ec96e9bc04862405da9/packages/block-library/src/table/state.js#L221-L244

It'll be an easy refactor from this to that.

Reply:
That's fine, but then I wouldn't call it `isCellSelected`? That's misleading if you're checking if a column is selected. `isCellSelected( ..., 'column' )` doesn't make sense?

Reply:
Hmm, yeah, that's good feedback. I can work on a few naming improvements in future PRs as this functionality develops. Something like `isCellWithinSelection` might be more appropriate.

Thanks for the reviews btw!